src/expressions/operator-expr.md

```diff
@@ -478,15 +478,48 @@ unsafe {
 assert_eq!(values[1], 3);
 ```
 
-#### Slice DST pointer to pointer cast
+#### Sized to sized pointer casts
 
-For slice types like `[T]` and `[U]`, the raw pointer types `*const [T]`, `*mut [T]`,
-`*const [U]`, and `*mut [U]` encode the number of elements in this slice. Casts between
-these raw pointer types preserve the number of elements. Note that, as a consequence,
-such casts do *not* necessarily preserve the size of the pointer's referent (e.g.,
-casting `*const [u16]` to `*const [u8]` will result in a raw pointer which refers to an
-object of half the size of the original). The same holds for `str` and any compound type
-whose unsized tail is a slice type, such as struct `Foo(i32, [u8])` or `(u64, Foo)`.
+A sized pointer to pointer cast `*const T as *const U` or `*mut T as *mut U`
+where `T: Sized` and `U: Sized` produces a value identical to the original
+pointer.
+
+#### Unsized to unsized pointer casts
+
+Any unsized to unsized pointer cast `*const T as *const U` or `*mut T as *mut U`
+produces a value identical to the original pointer if `T` and `U` are compatible
+unsized types.
+
+For slice types like `[T]` and `[U]`, the raw pointer types `*const [T]`, `*mut
+[T]`, `*const [U]`, and `*mut [U]` encode the number of elements in this slice.
+Casts between these raw pointer types preserve the number of elements. Note
+that, as a consequence, such casts do *not* necessarily preserve the size of the
+pointer's referent (e.g., casting `*const [u16]` to `*const [u8]` will result in
+a raw pointer which refers to an object of half the size of the original). The
+same holds for `str` and any compound type whose unsized tail is a slice type,
+such as struct `Foo(i32, [u8])` or `(u64, Foo)`.
+
+#### Unsized to sized pointer casts
+
+Any wide to narrow pointer cast discards all metadata that completes the wide
+pointer and produces a thin base pointer.
+
+The following casts are well-defined:
+
+* `*const [T] to *const U` or `*mut [T] to *mut U` discards the length of the
+  slice and produces a pointer to the first element. The produced pointer is
+  casted to `U` in the same manner as a sized pointer to pointer cast.
+* `*const dyn T to *const U` or `*mut dyn T to *mut U` discards all metadata
+  associated with the [trait object] and produces a pointer to the instance from
+  which the trait object was constructed. The produced pointer is casted to `U`
+  in the same manner as a sized pointer to pointer cast.
```

Comment:
This list is not exhaustive, e.g. it misses this case:
```rust
fn test(x: *const (u8, str)) -> *const u16 { x as _ }
```
Also, it seems strange to list both of these only to then say they do basically the same. The enumeration just kind of repeats what the text above says?

Reply:
So you either want to also enumerate unsized tuples and presumably unsized custom types or collapse all of them into one?

It would be nice to list the common instances (`*const [T]` / `*const dyn T`) somehow, but maybe not necessary?

I would like to defer to your preference on this matter, anything works for me as long as the unsized pointer cast semantics are documented.

Reply:
Ah, saw the last comment now.

Reply:
> It would be nice to list the common instances (*const [T] / *const dyn T) somehow, but maybe not necessary?

This is a reference, not tutorial material. So no, common cases should not be listed. Every special case listed should only be listed because it actually is a special case that behaves different from the general case.

Reply:
Fair. If you feel it's useful, add something like that -- but the "for example" wording is crucial here.

However, I'm not convinced that an example is actually helpful here. This will be easier to discuss once there is concrete text for the revised version. :)